### PR TITLE
Add tests for path evals and a few new evals

### PR DESCRIPTION
**Pull request by valedan**
Changes:
- Unit tests for all evals
- Removed `len(prediction) <= 1` check as I don't think it is correct behaviour (pending discussion with Micheal on this one)
- Added `exact_path_predicted`, `solution_length`, `streak_length_until_incorrect` evals
- Ensured all evals return floats (not sure this is necessary, but seems best to be consistent in the interface)

### CODE REVIEW

File: maze_transformer/evaluation/path_evals.py [hunk 1]
```diff
@@ -41,29 +45,20 @@ class PathEvals:
     @register_method(evals)
     @staticmethod
     def node_overlap(solution: MazePath, prediction: MazePath, **_) -> float:
-        """number of shared nodes (any order) / total number of (unique) nodes"""
-        if len(prediction) <= 1:
```
Comment (mivanit) [inline maze_transformer/evaluation/path_evals.py:45]: yeah, not sure why this was even here

File: maze_transformer/evaluation/path_evals.py [hunk 2]
```diff
@@ -41,29 +45,20 @@ class PathEvals:
...
 
-        n_shared: int = 0
         solution_set = {tuple(coord) for coord in solution}
         prediction_set = {tuple(coord) for coord in prediction}
 
-        for coord in solution:
-            if tuple(coord) in prediction_set:
-                n_shared += 1
-        return n_shared / len(solution_set)
+        return len(prediction_set & solution_set) / len(solution_set)
```
Comment (mivanit) [inline maze_transformer/evaluation/path_evals.py:53]: very clean!

File: maze_transformer/evaluation/path_evals.py [hunk 3]
```diff
@@ -41,29 +45,20 @@ class PathEvals:
...
 
     @register_method(evals)
     @staticmethod
     def num_connections_adjacent_lattice(prediction: MazePath, **_) -> float:
         """number of the connections in prediction which actually connect nodes that are adjacent on the lattice, ignoring if they are adjacent on the maze"""
-        if len(prediction) <= 1:
-            return 0.0
-
-        n_adj: int = 0
+        n_adj = 0.0
```
Comment (mivanit) [inline maze_transformer/evaluation/path_evals.py:59]: no type hints :'(

This is definitely a nitpick but I think it helps us with making mypy checking be useful down the line

File: maze_transformer/evaluation/path_evals.py [hunk 4]
```diff
@@ -97,4 +88,40 @@ def fraction_connections_adjacent(
...
 
-        return PathEvals.num_connections_adjacent(maze, prediction) / len(prediction)
+        num_connections = len(prediction) - 1
+        return PathEvals.num_connections_adjacent(maze, prediction) / num_connections
+
+    @register_method(evals)
+    @staticmethod
+    def exact_path_predicted(solution: MazePath, prediction: MazePath, **_) -> float:
+        """Was the maze successfully solved?"""
+        return float(np.all(solution == prediction))
```
Comment (mivanit) [inline maze_transformer/evaluation/path_evals.py:98]: I would add a comment here explaining that `==` doesn't actually crash for arrays of different sizes -- I was certainly under the impression that they do

**Review by mivanit — APPROVED** [review]
LGTM! Only real comment is that I would maybe add some type hints.

**Comment by valedan** [comment]
Type inference without a hint - hovering `n_adj` shows the correct type:

Screenshot 2023-04-07 at 11 56 03 AM

Does this work on your editor?

**Comment by mivanit** [comment]
> @mivanit Re: type hints - I might be misunderstanding something, but I thought I put type hints everywhere they were needed (ie all function definitions and anywhere that mypy can't infer types).
> 
> For the example you pointed out above - `n_adj: int = 0` I believe the type hint is not necessary here because mypy correctly [infers](https://mypy.readthedocs.io/en/stable/type_inference_and_annotations.html#type-inference) the type without it, and the hint serves no documentation purpose because it's obvious to any reader what is going on.
> 
> Is there a reason I'm missing for having that specific type hint, or are there other places I should have put some?

I always try to add type hints when defining a variable, since this also serves the added benefit of distinguishing between definitions and assignments for any given variable. Even if the actual type itself is obvious, and the code is simple enough to where no documentation is needed, I think it helps to be consistent in "always type hint on var defn" rather than deciding on a case-by-case basis when it is or is not needed.

Honestly, though, this comes from a place of me being neurotic and not a place of logic. Maybe I subconsciously just miss C++ :')

**Comment by valedan** [comment]
Ah okay fair! I think I personally lean a bit more minimalist with this stuff and like to add hints and comments etc only when necessary, but codebase consistency is more important so I'll add them back in. 🙂

File: maze_transformer/evaluation/path_evals.py (hunk 4 above)
Comment (valedan) [reply maze_transformer/evaluation/path_evals.py:98]: This is actually a good catch! It does work (numpy broadcasts), but it is deprecated behaviour and emits a warning. Fixing.

**Pull request merged**